### PR TITLE
API: optionally use a user supplied preimage when creating invoices

### PR DESCRIPTION
**Pull request by araspitzu**
This enriches the `/createinvoice` API to accept a custom preimage and allows the caller to specify the preimage that will be used to generate the payment hash and invoice, should be used with care (low entropy on the preimage might get your incoming payment stolen).  This PR builds on top of #938 .

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala [hunk 1]
```diff
@@ -57,7 +58,7 @@ trait Eclair {
 
   def peersInfo()(implicit timeout: Timeout): Future[Iterable[PeerInfo]]
 
-  def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String])(implicit timeout: Timeout): Future[PaymentRequest]
+  def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], preimage:Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest]
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala]:
```suggestion
  def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], paymentPreimage: Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest]
```

File: eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala [hunk 2]
```diff
@@ -152,9 +153,9 @@ class EclairImpl(appKit: Kit) extends Eclair {
     case Some(pk) => (appKit.router ? 'updatesMap).mapTo[Map[ChannelDesc, ChannelUpdate]].map(_.filter(e => e._1.a == pk || e._1.b == pk).values)
   }
 
-  override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String])(implicit timeout: Timeout): Future[PaymentRequest] = {
+  override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], preimage:Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest] = {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala]:
```suggestion
  override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], paymentPreimage: Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest] = {
```

File: eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala [hunk 3]
```diff
@@ -152,9 +153,9 @@ class EclairImpl(appKit: Kit) extends Eclair {
...
     fallbackAddress.map { fa => fr.acinq.eclair.addressToPublicKeyScript(fa, appKit.nodeParams.chainHash) } // if it's not a bitcoin address throws an exception
-    (appKit.paymentHandler ? ReceivePayment(description = description, amountMsat_opt = amountMsat.map(MilliSatoshi), expirySeconds_opt = expire, fallbackAddress = fallbackAddress)).mapTo[PaymentRequest]
+    (appKit.paymentHandler ? ReceivePayment(description = description, amountMsat_opt = amountMsat.map(MilliSatoshi), expirySeconds_opt = expire, fallbackAddress = fallbackAddress, preimage = preimage)).mapTo[PaymentRequest]
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/Eclair.scala]:
```suggestion
    (appKit.paymentHandler ? ReceivePayment(description = description, amountMsat_opt = amountMsat.map(MilliSatoshi), expirySeconds_opt = expire, fallbackAddress = fallbackAddress, paymentPreimage = paymentPreimage)).mapTo[PaymentRequest]
```

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 4]
```diff
@@ -229,8 +229,8 @@ trait Service extends ExtraDirectives with Logging {
                         }
                       } ~
                       path("createinvoice") {
-                        formFields("description".as[String], amountMsatFormParam.?, "expireIn".as[Long].?, "fallbackAddress".as[String].?) { (desc, amountMsat, expire, fallBackAddress) =>
-                          complete(eclairApi.receive(desc, amountMsat, expire, fallBackAddress))
+                        formFields("description".as[String], amountMsatFormParam.?, "expireIn".as[Long].?, "fallbackAddress".as[String].?, "preimage".as[ByteVector32](sha256HashUnmarshaller).?) { (desc, amountMsat, expire, fallBackAddress, preimage_opt) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]:
```suggestion
                        formFields("description".as[String], amountMsatFormParam.?, "expireIn".as[Long].?, "fallbackAddress".as[String].?, "paymentPreimage".as[ByteVector32](sha256HashUnmarshaller).?) { (desc, amountMsat, expire, fallBackAddress, paymentPreimage_opt) =>
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala [hunk 5]
```diff
@@ -43,9 +43,9 @@ class LocalPaymentHandler(nodeParams: NodeParams) extends Actor with ActorLoggin
 
   override def receive: Receive = {
 
-    case ReceivePayment(amount_opt, desc, expirySeconds_opt, extraHops, fallbackAddress_opt) =>
+    case ReceivePayment(amount_opt, desc, expirySeconds_opt, extraHops, fallbackAddress_opt, preimage_opt) =>
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]:
```suggestion
    case ReceivePayment(amount_opt, desc, expirySeconds_opt, extraHops, fallbackAddress_opt, paymentPreimage_opt) =>
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala [hunk 6]
```diff
@@ -43,9 +43,9 @@ class LocalPaymentHandler(nodeParams: NodeParams) extends Actor with ActorLoggin
...
       Try {
-        val paymentPreimage = randomBytes32
+        val paymentPreimage = preimage_opt.getOrElse(randomBytes32)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/LocalPaymentHandler.scala]:
```suggestion
        val paymentPreimage = paymentPreimage_opt.getOrElse(randomBytes32)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala [hunk 7]
```diff
@@ -191,7 +191,7 @@ object PaymentLifecycle {
   def props(nodeParams: NodeParams, id: UUID, router: ActorRef, register: ActorRef) = Props(classOf[PaymentLifecycle], nodeParams, id, router, register)
 
   // @formatter:off
-  case class ReceivePayment(amountMsat_opt: Option[MilliSatoshi], description: String, expirySeconds_opt: Option[Long] = None, extraHops: List[List[ExtraHop]] = Nil, fallbackAddress: Option[String] = None)
+  case class ReceivePayment(amountMsat_opt: Option[MilliSatoshi], description: String, expirySeconds_opt: Option[Long] = None, extraHops: List[List[ExtraHop]] = Nil, fallbackAddress: Option[String] = None, preimage:Option[ByteVector32] = None)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentLifecycle.scala]:
```suggestion
  case class ReceivePayment(amountMsat_opt: Option[MilliSatoshi], description: String, expirySeconds_opt: Option[Long] = None, extraHops: List[List[ExtraHop]] = Nil, fallbackAddress: Option[String] = None, paymentPreimage: Option[ByteVector32] = None)
```

File: eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala [hunk 8]
```diff
@@ -153,4 +157,21 @@ class EclairImplSpec extends TestKit(ActorSystem("mySystem")) with fixture.FunSu
     register.expectMsg(Register.ForwardShortId(ShortChannelId("568749x2597x0"), CMD_FORCECLOSE))
   }
 
+  test("passing a preimage to /createinvoice should result in an invoice with payment_hash=H(preimage)") { fixture =>
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala]:
```suggestion
  test("passing a payment_preimage to /createinvoice should result in an invoice with payment_hash=H(payment_preimage)") { fixture =>
```

File: eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala [hunk 9]
```diff
@@ -153,4 +157,21 @@ class EclairImplSpec extends TestKit(ActorSystem("mySystem")) with fixture.FunSu
...
+
+    val paymentHandler = system.actorOf(LocalPaymentHandler.props(Alice.nodeParams))
+    val kitWithPaymentHandler = fixture.kit.copy(paymentHandler = paymentHandler)
+    val eclair = new EclairImpl(kitWithPaymentHandler)
+    val preimage = randomBytes32
+
+    val fResp = eclair.receive(description = "some desc", amountMsat = None, expire = None, fallbackAddress = None, preimage = Some(preimage))
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala]:
```suggestion
    val fResp = eclair.receive(description = "some desc", amountMsat = None, expire = None, fallbackAddress = None, paymentPreimage = Some(paymentPreimage))
```

File: eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala [hunk 10]
```diff
@@ -153,4 +157,21 @@ class EclairImplSpec extends TestKit(ActorSystem("mySystem")) with fixture.FunSu
     register.expectMsg(Register.ForwardShortId(ShortChannelId("568749x2597x0"), CMD_FORCECLOSE))
   }
 
+  test("passing a preimage to /createinvoice should result in an invoice with payment_hash=H(preimage)") { fixture =>
+
+    val paymentHandler = system.actorOf(LocalPaymentHandler.props(Alice.nodeParams))
+    val kitWithPaymentHandler = fixture.kit.copy(paymentHandler = paymentHandler)
+    val eclair = new EclairImpl(kitWithPaymentHandler)
+    val preimage = randomBytes32
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/EclairImplSpec.scala]:
```suggestion
    val paymentPreimage = randomBytes32
```

File: eclair-core/src/test/scala/fr/acinq/eclair/api/ApiServiceSpec.scala [hunk 11]
```diff
@@ -65,7 +65,7 @@ class ApiServiceSpec extends FunSuite with ScalatestRouteTest {
 
     override def peersInfo()(implicit timeout: Timeout): Future[Iterable[PeerInfo]] = ???
 
-    override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String])(implicit timeout: Timeout): Future[PaymentRequest] = ???
+    override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], preimage: Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest] = ???
```
Comment (araspitzu) [inline eclair-core/src/test/scala/fr/acinq/eclair/api/ApiServiceSpec.scala]:
```suggestion
    override def receive(description: String, amountMsat: Option[Long], expire: Option[Long], fallbackAddress: Option[String], paymentPreimage: Option[ByteVector32])(implicit timeout: Timeout): Future[PaymentRequest] = ???
```

**Comment by araspitzu** [comment]
TODO: update doc

**Pull request merged**